File: 0901-1000/0902-Numbers At Most N Given Digit Set/0902-Numbers At Most N Given Digit Set.cpp
class Solution {
public:
    int atMostNGivenDigitSet(vector<string>& D, int N) {
        string s = to_string(N);
        int n = s.size();
        vector<int> dp(1 + n);
        dp[0] = 1;
        for (int i = 1; i <= n; ++i) {
            for (int j = D.size() - 1; j >= 0; --j) {
                if (D[j][0] == s[n - i]) {
                    dp[i] = dp[i - 1] + j * pow(D.size(), i - 1);
                    break;
                }
                else if (D[j][0] < s[n - i]) {
                    dp[i] = (j + 1) * pow(D.size(), i - 1);
                    break;
                }
            }
        }
        int total = dp[n];
        for (int i = 1; i < n; ++i) {
            total += pow(D.size(), i);
        }
        return total;
    }
};

File: 0901-1000/0902-Numbers At Most N Given Digit Set/0902-Numbers At Most N Given Digit Set.py
class Solution:
    def atMostNGivenDigitSet(self, D: List[str], N: int) -> int:
        s = str(N)
        # dp[i] means number with length i from right to left
        dp = [0] * (len(s) + 1)
        dp[0] = 1
        for i in range(1, len(s) + 1):
            for j in range(len(D) - 1, -1, -1):
                if D[j] == s[-i]:
                    dp[i] = dp[i - 1] + j * (len(D) ** (i - 1))
                    break
                elif D[j] < s[-i]:
                    dp[i] = (j + 1) * (len(D) ** (i - 1))
                    break
        
        return dp[-1] + sum(len(D) ** i for i in range(1, len(s)))
